### PR TITLE
Audit trail

### PR DESCRIPTION
**Pull request by weezyjr**
### 1- Header: Audit trail: Make sure that you can export all the fields in the active table ... 	
```
2019-12-24 16:25:42.286 | INFO     | ui_testing.testcases.base_test:setUp:25 - Test case : test001_download_audit_trail_sheet
2019-12-24 16:25:43.742 | INFO     | ui_testing.pages.login_page:login:10 - Login admin : admin.
2019-12-24 16:26:15.767 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2019-12-24 16:26:20.773 | INFO     | ui_testing.pages.base_pages:info:229 -  + Download XSLX sheet
2019-12-24 16:26:52.114 | INFO     | ui_testing.pages.base_pages:info:229 -  + Comparing the audit trail no. 0 
2019-12-24 16:26:52.179 | INFO     | ui_testing.testcases.base_test:tearDown:42 - TearDown. 	
FAIL
```
### 2- Header: Audit trail Approach: Make sure that I can filter by all the following fields ... 	
```
2019-12-24 16:26:52.180 | INFO     | ui_testing.testcases.base_test:setUp:25 - Test case : test002_audit_trail_filters
2019-12-24 16:26:53.613 | INFO     | ui_testing.pages.login_page:login:10 - Login admin : admin.
2019-12-24 16:27:37.040 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2019-12-24 16:27:42.364 | INFO     | ui_testing.pages.base_pages:open_filter_menu:69 -  Open Filter
2019-12-24 16:27:42.488 | INFO     | ui_testing.pages.audit_trail_page:filter_audit_trail_by:21 -  + Filter by test plan : 24.12.2019
2019-12-24 16:27:48.218 | INFO     | ui_testing.pages.base_pages:sleep_tiny:35 -  Tiny sleep.
2019-12-24 16:27:50.883 | INFO     | ui_testing.pages.audit_trail_page:filter_audit_trail_by:21 -  + Filter by test plan : admin
2019-12-24 16:27:58.559 | INFO     | ui_testing.pages.base_pages:sleep_tiny:35 -  Tiny sleep.
2019-12-24 16:28:01.258 | INFO     | ui_testing.pages.audit_trail_page:filter_audit_trail_by:21 -  + Filter by test plan : Archive
2019-12-24 16:28:08.776 | INFO     | ui_testing.pages.base_pages:sleep_tiny:35 -  Tiny sleep.
2019-12-24 16:28:11.469 | INFO     | ui_testing.pages.audit_trail_page:filter_audit_trail_by:21 -  + Filter by test plan : Article
2019-12-24 16:28:18.976 | INFO     | ui_testing.pages.base_pages:sleep_tiny:35 -  Tiny sleep.
2019-12-24 16:28:21.639 | INFO     | ui_testing.pages.audit_trail_page:filter_audit_trail_by:21 -  + Filter by test plan : 379'470'239
2019-12-24 16:28:27.092 | INFO     | ui_testing.pages.base_pages:sleep_tiny:35 -  Tiny sleep.
2019-12-24 16:28:29.195 | INFO     | ui_testing.pages.base_pages:open_filter_menu:69 -  Open Filter
2019-12-24 16:28:29.334 | INFO     | ui_testing.testcases.base_test:tearDown:42 - TearDown. 	
ok
```
### 3- Header: Audit trail: Approach: Make sure that you can search by all the followings entities ... 	
```
2019-12-24 16:28:29.335 | INFO     | ui_testing.testcases.base_test:setUp:25 - Test case : test003_search_audit_trail
2019-12-24 16:28:30.911 | INFO     | ui_testing.pages.login_page:login:10 - Login admin : admin.
2019-12-24 16:29:04.732 | INFO     | ui_testing.pages.base_pages:sleep_small:39 -  Small sleep.
2019-12-24 16:29:10.423 | INFO     | ui_testing.pages.base_pages:sleep_medium:43 -  Medium sleep.
2019-12-24 16:29:21.432 | INFO     | ui_testing.pages.base_pages:sleep_medium:43 -  Medium sleep.
2019-12-24 16:29:32.492 | INFO     | ui_testing.pages.base_pages:sleep_medium:43 -  Medium sleep.
2019-12-24 16:29:43.398 | INFO     | ui_testing.pages.base_pages:sleep_medium:43 -  Medium sleep.
2019-12-24 16:29:54.153 | INFO     | ui_testing.testcases.base_test:tearDown:42 - TearDown. 	
ok
```

### CODE REVIEW

**Comment by weezyjr** [comment]
https://modeso.atlassian.net/browse/LIMS-6399
Created bug for the first test case

File: ui_testing/pages/audit_trail_page.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        self.audit_trails_url = "{}auditTrails".format(self.base_selenium.url)
+
+    def get_audit_trails_page(self):
+        self.base_selenium.get(url=self.audit_trails_url)
+        self.sleep_small()
+
+    def download_xslx_sheet(self):
+        self.base_selenium.scroll()
+        self.base_selenium.click(element='general:right_menu')
+        self.sheet = self.base_selenium.download_excel_file(element='general:xslx')
```
Comment (0xIslamTaha) [inline ui_testing/pages/audit_trail_page.py]: already in BasePage

File: ui_testing/pages/audit_trail_page.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    def get_table_rows_data(self):
+        return [row.text for row in self.base_selenium.get_table_rows(element='general:table')]
```
Comment (0xIslamTaha) [inline ui_testing/pages/audit_trail_page.py]: already in basepage

File: ui_testing/pages/audit_trail_page.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    def filter_audit_trail_by(self, filter_name , filter_text, field_type='drop_down'):
+        self.base_selenium.LOGGER.info(' + Filter by test plan : {}'.format(filter_text))
+        self.filter_by(filter_element='audit_trail:filter_{}'.format(filter_name), filter_text=filter_text, field_type=field_type)
+        self.filter_apply()
+        self.sleep_tiny()
+        return self.get_table_rows_data()[0]
+
+    def get_random_audit_trail_row(self):
+         return self.get_random_table_row(table_element='general:table')
```
Comment (0xIslamTaha) [inline ui_testing/pages/audit_trail_page.py]: no need to create a method which returns another method

File: ui_testing/pages/audit_trail_page.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    def search(self, value):
+        """
+        Search for a specific value
+        :param value:
+        :return: The first element in the search table
+        """
+        self.base_selenium.set_text(element='general:search', value=value)
+        self.base_selenium.click(element='general:search')
+        self.sleep_medium()
+        return self.get_table_rows_data()[0]
```
Comment (0xIslamTaha) [inline ui_testing/pages/audit_trail_page.py]: already in BasePage

File: ui_testing/testcases/base_test.py [hunk 5]
```diff
@@ -49,7 +51,10 @@ def fix_data_format(self, data_list):
         tmp = []
         for item in data_list:
             if len(str(item)) > 0:
-                if re.search(r'\d{2}.\d{2}.\d{4}', str(item)):
+                if re.search(r'\d{2}.\d{2}.\d{4},\s\d{1,2}:\d{1,2}\s(A|P)M', str(item)):                    
+                    tmp.append(datetime.datetime.strptime(item, '%d.%m.%Y, %H:%M %p'))
+                elif re.search(r'\d{2}.\d{2}.\d{4}', str(item)):
+                    # item = item.split(',')[0] # if the string have time, remove the time
```
Comment (0xIslamTaha) [inline ui_testing/testcases/base_test.py]: this gonna affect other test cases, Are u sure the test case using it is passing? if not, remove this change and create your own method

File: ui_testing/testcases/basic_tests/test06_audit_trail.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+from parameterized import parameterized
+import re
+
+class AuditTrailTestCases(BaseTest):
+    def setUp(self):
+        super().setUp()
+        self.login_page.login(username=self.base_selenium.username, password=self.base_selenium.password)
+        self.base_selenium.wait_until_page_url_has(text='dashboard')
+        self.audit_trail_page.get_audit_trails_page()
+
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_audit_trail.py]:
```suggestion
@skip("WITH BUG ID")
```

File: ui_testing/testcases/basic_tests/test06_audit_trail.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        self.audit_trail_page.download_xslx_sheet()
+        rows_data = self.audit_trail_page.get_table_rows_data()
+        for index in range(len(rows_data)):
+            self.audit_trail_page.info(' + Comparing the audit trail no. {} '.format(index))
+            fixed_row_data = self.fix_data_format(rows_data[index].split('\n'))
+            values = self.audit_trail_page.sheet.iloc[index].values
+            fixed_sheet_row_data = self.fix_data_format(values)
+            for item in fixed_row_data:
+                self.assertIn(item, fixed_sheet_row_data)
+        pass
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_audit_trail.py]: remove

File: ui_testing/testcases/basic_tests/test06_audit_trail.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        result = self.audit_trail_page.filter_audit_trail_by(filter_name='entity', filter_text=entity, field_type='drop_down')
+        self.assertIn(entity, result)
+
+        # filter by Change By
+        result = self.audit_trail_page.filter_audit_trail_by(filter_name='entity_number', filter_text=entity_number, field_type='text')
+        self.assertIn(entity_number, result)
+        
+        # close filter menu
+        self.audit_trail_page.open_filter_menu()
+        pass
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_audit_trail.py]: remove

File: ui_testing/testcases/basic_tests/test06_audit_trail.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        audit_trail_row = self.audit_trail_page.get_random_audit_trail_row()
+        # map the header to the row data
+        audit_trail_row_data = self.base_selenium.get_row_cells_dict_related_to_header(row=audit_trail_row)
+        
+        # get each column data from that row
+        action_date = audit_trail_row_data['Action Date'].split(',')[0] # get only the date
+        changed_by = audit_trail_row_data['Changed By']
+        action = audit_trail_row_data['Action']
+        entity = audit_trail_row_data['Entity']
+        entity_number = audit_trail_row_data['Entity Number']
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_audit_trail.py]: create a method with those lines and use it in this test case and the next text case

File: ui_testing/testcases/basic_tests/test06_audit_trail.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+            fixed_sheet_row_data = self.fix_data_format(values)
+            for item in fixed_row_data:
+                self.assertIn(item, fixed_sheet_row_data)
+        pass    
+                
+    
+    def test002_audit_trail_filters(self):
+        """
+        Header: Audit trail Approach: Make sure that I can filter by all the following fields 
+        ( action date & entity & action & entity no & changed by )
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_audit_trail.py]: each one should be a test case. Separate them

File: ui_testing/testcases/basic_tests/test06_audit_trail.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        self.assertIn(action, result)
+
+        # search by entity
+        result = self.audit_trail_page.search(entity)
+        self.assertIn(entity, result)
+
+        # search by entity number
+        result = self.audit_trail_page.search(entity_number)
+        self.assertIn(entity_number, result)
+        pass
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_audit_trail.py]: remove

File: ui_testing/testcases/basic_tests/test06_audit_trail.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        # get random row in the table
+        audit_trail_row = self.audit_trail_page.get_random_audit_trail_row()
+        # map the header to the row data
+        audit_trail_row_data = self.base_selenium.get_row_cells_dict_related_to_header(row=audit_trail_row)
+        
+        # get each column data from that row
+        changed_by = audit_trail_row_data['Changed By']
+        action = audit_trail_row_data['Action']
+        entity = audit_trail_row_data['Entity']
+        entity_number = audit_trail_row_data['Entity Number']
```
Comment (0xIslamTaha) [inline ui_testing/testcases/basic_tests/test06_audit_trail.py]: use a method

File: ui_testing/testcases/basic_tests/test06_audit_trail.py (hunk 11 above)
Comment (weezyjr) [reply ui_testing/testcases/basic_tests/test06_audit_trail.py]: @0xIslamTaha  why ?

File: ui_testing/testcases/base_test.py (hunk 5 above)
Comment (weezyjr) [reply ui_testing/testcases/base_test.py]: @0xIslamTaha  No, it won't affect any other test cases

File: ui_testing/testcases/basic_tests/test06_audit_trail.py (hunk 7 above)
Comment (weezyjr) [reply ui_testing/testcases/basic_tests/test06_audit_trail.py]: @0xIslamTaha why ?

**Pull request merged**